src/core/vm.c

```diff
@@ -350,9 +350,11 @@ __attribute__((weak)) cpumap_t vm_translate_to_vcpu_mask(struct vm* vm, cpumap_t
 
 void vcpu_run(struct vcpu* vcpu)
 {
-    if (vcpu->active == false) {
-        cpu_idle();
-    } else {
-        vcpu_arch_run(vcpu);
+    if (vcpu_arch_is_on(vcpu) && vcpu->active) {
+        vcpu_arch_entry();
+    } else if (!vcpu_arch_is_on(vcpu)) {
+        cpu_powerdown();
+    } else if (vcpu_arch_is_on(vcpu) && !vcpu->active) {
+        cpu_standby();
```

Comment:
This flow is a bit confusing IMO

Please re-write it like this:
```
void vcpu_run(struct vcpu* vcpu)
{
    if (vcpu_arch_is_on(vcpu)){
        if(vcpu->active) {
            vcpu_arch_entry();
        } else  {
            cpu_standby();
        }
    } else {
        cpu_powerdown();
    }
}
```

Reply:
> This flow is a bit confusing IMO
> 
> Please re-write it like this:
> 
> ```
> void vcpu_run(struct vcpu* vcpu)
> {
>     if (vcpu_arch_is_on(vcpu)){
>         if(vcpu->active) {
>             vcpu_arch_entry();
>         } else  {
>             cpu_standby();
>         }
>     } else {
>         cpu_powerdown();
>     }
> }
> ```

Done [here](https://github.com/bao-project/bao-hypervisor/pull/186/commits/8adff2474cd9cabb4b61c1bf40482538571af909)